File: src/core/telemetry.js
export class Telemetry {
    constructor(state) {
        this.state = state;
        this.live = new Map();
        this.listeners = [];
    }

    onEvent(callback) {
        this.listeners.push(callback);
    }

    _broadcast(event) {
        for (const cb of this.listeners) {
            try { cb(event); } catch {}
        }
    }

    connectorStart(investigationId, connector, phaseId, input) {
        const event = {
            type: 'connector_start',
            investigationId,
            connector,
            phaseId,
            input,
            timestamp: Date.now()
        };
        this.live.set(`${investigationId}:${connector}:${input?.value}`, event);
        this.state.logTelemetry(investigationId, 'connector_start', event, connector, phaseId);
        this._broadcast(event);
    }

    connectorEnd(investigationId, connector, phaseId, result) {
        const key = `${investigationId}:${connector}:${result?.input?.value}`;
        const start = this.live.get(key);
        const duration = start ? Date.now() - start.timestamp : 0;

        const event = {
            type: 'connector_end',
            investigationId,
            connector,
            phaseId,
            status: result.status,
            entitiesFound: result.entitiesFound || 0,
            input: result.input,
            duration,
            timestamp: Date.now()
        };
        this.live.delete(key);
        this.state.logTelemetry(investigationId, 'connector_end', event, connector, phaseId);
        this._broadcast(event);
    }

    phaseStart(investigationId, phaseId) {
        const event = { type: 'phase_start', investigationId, phaseId, timestamp: Date.now() };
        this.state.logTelemetry(investigationId, 'phase_start', event, null, phaseId);
        this._broadcast(event);
    }

    phaseEnd(investigationId, phaseId, status) {
        const event = { type: 'phase_end', investigationId, phaseId, status, timestamp: Date.now() };
        this.state.logTelemetry(investigationId, 'phase_end', event, null, phaseId);
        this._broadcast(event);
    }

    entityDiscovered(investigationId, entityType, isNew, connector) {
        const event = {
            type: isNew ? 'entity_new' : 'entity_updated',
            investigationId,
            entityType,
            connector,
            timestamp: Date.now()
        };
        this.state.logTelemetry(investigationId, event.type, event, connector);
        this._broadcast(event);
    }

    error(investigationId, connector, message, phaseId = null) {
        const event = {
            type: 'error',
            investigationId,
            connector,
            message,
            phaseId,
            timestamp: Date.now()
        };
        this.state.logTelemetry(investigationId, 'error', event, connector, phaseId);
        this._broadcast(event);
    }

    getActiveRuns() {
        return Array.from(this.live.values());
    }

    getSummary(investigationId) {
        return this.state.getStats(investigationId);
    }
}

File: cyclops/core/telemetry.py
from datetime import datetime, timezone


class Telemetry:
    def __init__(self):
        self._listeners = []
        self._active_runs = {}

    def on_event(self, callback):
        self._listeners.append(callback)

    def _emit(self, event):
        event["timestamp"] = datetime.now(timezone.utc).isoformat()
        for cb in self._listeners:
            try:
                cb(event)
            except Exception:
                pass

    def phase_start(self, inv_id, phase_id):
        self._emit({"type": "phase_start", "investigationId": inv_id, "phaseId": phase_id})

    def phase_end(self, inv_id, phase_id, status):
        self._emit({"type": "phase_end", "investigationId": inv_id, "phaseId": phase_id, "status": status})

    def connector_start(self, inv_id, connector, phase_id, meta=None):
        key = f"{inv_id}:{connector}:{meta.get('value', '') if meta else ''}"
        self._active_runs[key] = {"connector": connector, "phase_id": phase_id, "meta": meta}
        self._emit({"type": "connector_start", "investigationId": inv_id, "connector": connector, "phaseId": phase_id, **(meta or {})})

    def connector_end(self, inv_id, connector, phase_id, meta=None):
        key = f"{inv_id}:{connector}:{meta.get('input', {}).get('value', '') if meta else ''}"
        self._active_runs.pop(key, None)
        self._emit({"type": "connector_end", "investigationId": inv_id, "connector": connector, "phaseId": phase_id, **(meta or {})})

    def entity_discovered(self, inv_id, entity_type, is_new, source):
        event_type = "entity_new" if is_new else "entity_dup"
        self._emit({"type": event_type, "investigationId": inv_id, "entityType": entity_type, "source": source})

    def get_active_runs(self):
        return list(self._active_runs.values())
